src/common/token_bucket.c

```diff
@@ -57,55 +103,56 @@ token_bucket_rw_init(token_bucket_rw_t *bucket,
  */
 void
 token_bucket_rw_adjust(token_bucket_rw_t *bucket,
-                    uint32_t rate,
-                    uint32_t burst)
+                       uint32_t rate,
+                       uint32_t burst)
 {
-  tor_assert_nonfatal(rate > 0);
-  tor_assert_nonfatal(burst > 0);
-  if (burst > TOKEN_BUCKET_RW_MAX_BURST)
-    burst = TOKEN_BUCKET_RW_MAX_BURST;
-
-  bucket->rate = rate_per_sec_to_rate_per_step(rate);
-  bucket->burst = burst;
-  bucket->read_bucket = MIN(bucket->read_bucket, (int32_t)burst);
-  bucket->write_bucket = MIN(bucket->write_bucket, (int32_t)burst);
+  token_bucket_cfg_init(&bucket->cfg,
+                        rate_per_sec_to_rate_per_step(rate),
+                        burst);
+  token_bucket_raw_adjust(&bucket->read_bucket, &bucket->cfg);
+  token_bucket_raw_adjust(&bucket->read_bucket, &bucket->cfg);
 }
 
 /**
  * Reset <b>bucket</b> to be full, as of timestamp <b>now_ts</b>.
  */
 void
 token_bucket_rw_reset(token_bucket_rw_t *bucket,
-                   uint32_t now_ts)
+                      uint32_t now_ts)
 {
-  bucket->read_bucket = bucket->burst;
-  bucket->write_bucket = bucket->burst;
-  bucket->last_refilled_at_ts = now_ts;
+  token_bucket_raw_reset(&bucket->read_bucket, &bucket->stamp,
+                         &bucket->cfg, now_ts);
+  token_bucket_raw_reset(&bucket->write_bucket, &bucket->stamp,
+                         &bucket->cfg, now_ts);
 }
 
-/* Helper: see token_bucket_rw_refill */
-static int
-refill_single_bucket(int32_t *bucketptr,
-                     const uint32_t rate,
-                     const int32_t burst,
-                     const uint32_t elapsed_steps)
+/**
+ * Given an amount of <b>elapsed</b> time units, and a bucket configuration
+ * <b>cfg</b>, refill the level of <b>bucket</b> accordingly.  Note that the
+ * units of time in <b>elapsed</b> must correspond to those used to set the
+ * rate in <b>cfg</b>, or the result will be illogical.
+ */
+int
+token_bucket_raw_refill_steps(token_bucket_raw_t *bucket,
```

Comment:
This can be made static.

Reply:
No; it should be public, so that others can use token_bucket_raw.

Reply:
That is to say, the idea of the refactoring here is that the token_bucket_raw() functions can be used by outside code to build token buckets that aren't specifically tied to token_bucket_rw_t.